cpp/src/column/column_device_view.cu

```diff
@@ -69,48 +69,65 @@ column_device_view::column_device_view( column_view source, ptrdiff_t h_ptr, ptr
-      return p;
-    }
-    // First calculate the size of memory needed to hold the
-    // child columns. This is done by calling extent()
-    // for each of the children.
-    size_type size_bytes = 0;
-    for( size_type idx=0; idx < num_children; ++idx )
-      size_bytes += extent(source.child(idx));
-    // A buffer of CPU memory is allocated to hold the column_device_view
-    // objects. Once filled, the CPU memory is copied to device memory
-    // and then set into the d_children member pointer.
-    std::vector<int8_t> h_buffer(size_bytes);
-    auto h_start = h_buffer.data();
-    // Each column_device_view instance may have child objects that
-    // require setting some internal device pointers before being copied
-    // from CPU to device.
-    int8_t* d_start;
-    RMM_TRY(RMM_ALLOC(&d_start, size_bytes, stream));
-    std::unique_ptr<column_device_view, decltype(deleter)> p{
-        new column_device_view(source,reinterpret_cast<ptrdiff_t>(h_start),reinterpret_cast<ptrdiff_t>(d_start)), deleter};
-
-    // copy the CPU memory with all the children into device memory
-    CUDA_TRY(cudaMemcpyAsync(d_start, h_start, size_bytes,
-                              cudaMemcpyHostToDevice, stream));
-    p->_num_children = num_children;
-    p->d_children = reinterpret_cast<column_device_view*>(d_start);
-    CUDA_TRY(cudaStreamSynchronize(stream));
-    return p;
+std::unique_ptr<column_device_view, std::function<void(column_device_view*)>>
+column_device_view::create(column_view source, cudaStream_t stream) {
+
+  size_type num_children = source.num_children();
+  if (num_children == 0) {
+    // Can't use make_unique since the ctor is protected
+    return std::unique_ptr<column_device_view>(new column_device_view(source));
+  }
+
+  // First calculate the size of memory needed to hold the
+  // child columns. This is done by calling extent()
+  // for each of the children.
+  auto get_extent = thrust::make_transform_iterator(
+      thrust::make_counting_iterator(0),
+      [&source](auto i) { return extent(source.child(i)); });
+
+  std::size_t const child_storage_bytes =
+      std::accumulate(get_extent, get_extent + num_children, 0);
+
+  // A buffer of CPU memory is allocated to hold the column_device_view
+  // objects. Once filled, the CPU memory is copied to device memory
+  // and then set into the d_children member pointer.
+  std::vector<char> h_buffer(child_storage_bytes);
+  char* h_start = h_buffer.data();
+
+  // Each column_device_view instance may have child objects that
+  // require setting some internal device pointers before being copied
+  // from CPU to device.
+  rmm::device_buffer* const child_storage = new rmm::device_buffer(child_storage_bytes, stream);
+  char* d_start = static_cast<char*>(child_storage->data());
+
+  auto deleter = [child_storage](column_device_view * v){
+    v->destroy();
+    delete child_storage;
+  };
+
+  using Deleter = std::function<void(column_device_view*)>;
```

Comment:
On line 72 you have the same type in the return type. Worth moving up to use in both places?